experimental/traffic-portal/src/app/core/users/roles/table/roles-table.component.ts

```diff
@@ -0,0 +1,112 @@
+import { BehaviorSubject } from "rxjs";
+import type { ResponseRole } from "trafficops-types";
+
+import { UserService } from "src/app/api";
+import { CurrentUserService } from "src/app/shared/current-user/current-user.service";
+import type { ContextMenuActionEvent, ContextMenuItem } from "src/app/shared/generic-table/generic-table.component";
+import { NavigationService } from "src/app/shared/navigation/navigation.service";
+/**
+ * RolesTableComponent is the controller for the "Roles" table.
+ */
+@Component({
+	selector: "tp-roles",
+	styleUrls: ["./roles-table.component.scss"],
+	templateUrl: "./roles-table.component.html"
+})
+export class RolesTableComponent implements OnInit {
+	/** List of roles */
+	public roles: Promise<Array<ResponseRole>>;
+	constructor(private readonly route: ActivatedRoute, private readonly headerSvc: NavigationService,
+		private readonly api: UserService, public readonly auth: CurrentUserService) {
+		this.fuzzySubject = new BehaviorSubject<string>("");
+		this.roles = this.api.getRoles();
+		this.headerSvc.headerTitle.next("Roles");
+	}
+
+	/** Initializes table data, loading it from Traffic Ops. */
+	public ngOnInit(): void {
+		this.route.queryParamMap.subscribe(
+			m => {
+				const search = m.get("search");
+				if (search) {
+					this.fuzzControl.setValue(decodeURIComponent(search));
+					this.updateURL();
+				}
+			},
+			e => {
+				console.error("Failed to get query parameters:", e);
+			}
+		);
+	}
+
+	/** Definitions of the table's columns according to the ag-grid API */
+	public columnDefs = [
+		{
+			field: "name",
+			headerName: "Name"
+		},
+		{
+			field: "description",
+			headerName: "Description",
+		},
+		{
+			field: "lastUpdated",
+			headerName: "Last Updated"
+		}
+	];
+
+	/** Definitions for the context menu items (which act on augmented roles data). */
+	public contextMenuItems: Array<ContextMenuItem<ResponseRole>> = [
+		{
+			href: (selectedRow: ResponseRole): string => `${selectedRow.name}`,
+			name: "View Role"
+		},
+		{
+			href: (selectedRow: ResponseRole): string => `${selectedRow.name}`,
+			name: "Open in New Tab",
+			newTab: true
+		},
+		{
+			href: "/core/users",
+			name: "View Users",
+			queryParams: (selectedRow: ResponseRole): Params => ({role: selectedRow.name})
+		},
+	];
+
+	/** A subject that child components can subscribe to for access to the fuzzy search query text */
+	public fuzzySubject: BehaviorSubject<string>;
+
+	/** Form controller for the user search input. */
+	public fuzzControl = new FormControl<string>("");
+
+	/** Update the URL's 'search' query parameter for the user's search input. */
+	public updateURL(): void {
+		this.fuzzySubject.next(this.fuzzControl.value ?? "");
+	}
+
+	/**
+	 * Handles a context menu event.
+	 *
+	 * @param a The action selected from the context menu.
+	 */
+	public handleContextMenu(a: ContextMenuActionEvent<Readonly<ResponseRole>>): void {
+		console.log("action:", a);
+	}
```

Comment:
Since this doesn't do anything, I think we can do without it

Reply:
well, that's not an action though. The table has no context menu actions, they're all links, so this function would never be called - and even if it were it wouldn't do anything besides log a console line.

Reply:
ok, I see your point. I have a delete action in the roles-detail PR. So, for now, I will let this code be there since I added test cases for the same (below comment).

Reply:
alright, well, seems silly to ask you to take it out, then.